storage/src/mmr/bitmap.rs

```diff
@@ -0,0 +1,315 @@
+
+    /// A Merkle tree with each leaf representing DIGEST_SIZE*8 bits of the bitmap.
+    ///
+    /// When a chunk of DIGEST_SIZE*8 bits is accumulated by the bitmap, it is added to this tree.
+    /// Because leaf elements can be updated when bits in the bitmap are flipped, this tree, while
+    /// based on an MMR structure, is not an MMR but a Merkle tree.  The MMR structure results in
+    /// reduced update overhead for elements being appended or updated near the tip compared to a
+    /// more typical balanced Merkle tree.
+    mmr: Mmr<H>,
+}
+
+impl<H: CHasher> Default for Bitmap<H> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<H: CHasher> Bitmap<H> {
+    const CHUNK_SIZE: usize = H::Digest::SERIALIZED_LEN;
+
+    /// Return a new empty bitmap.
+    pub fn new() -> Self {
+        Bitmap {
+            bitmap: vec![0u8; H::Digest::SERIALIZED_LEN],
+            next_bit: 0,
+            mmr: Mmr::new(),
+        }
+    }
+
+    /// Return the number of bits currently stored in the bitmap.
+    pub fn bit_count(&self) -> u64 {
+        (self.bitmap.len() * 8 - Self::CHUNK_SIZE * 8 + self.next_bit) as u64
+    }
+
+    /// Return the last chunk of the bitmap as a slice.
+    fn last_chunk(&self) -> &[u8] {
+        let len = self.bitmap.len();
+        &self.bitmap[len - Self::CHUNK_SIZE..len]
+    }
+
+    /// Return the last chunk of the bitmap as a mutable slice.
+    fn last_chunk_mut(&mut self) -> &mut [u8] {
+        let len = self.bitmap.len();
+        &mut self.bitmap[len - Self::CHUNK_SIZE..len]
+    }
+
+    /// Commit the last chunk of the bitmap to the Merkle tree and initialize the next chunk.
+    fn commit_last_chunk(&mut self, hasher: &mut H) {
+        let chunk = H::Digest::try_from(self.last_chunk()).unwrap();
+        self.mmr.add(hasher, &chunk);
+        self.next_bit = 0;
+        self.bitmap.extend(vec![0u8; Self::CHUNK_SIZE]);
+    }
+
+    /// Efficiently add a digest-sized chunk of bits to the bitmap. Assumes we are at a chunk
+    /// boundary (that is, `self.next_bit` is 0) and panics otherwise.
+    pub fn append_chunk_unchecked(&mut self, hasher: &mut H, chunk: &H::Digest) {
+        assert!(
+            self.next_bit == 0,
+            "cannot add chunk when not chunk aligned"
+        );
+
+        self.last_chunk_mut().copy_from_slice(chunk.as_ref());
+        self.commit_last_chunk(hasher);
+    }
+
+    /// Efficiently add a byte's worth of bits to the bitmap. Assumes self.next_bit is currently
+    /// byte aligned, and panics otherwise.
+    pub fn append_byte_unchecked(&mut self, hasher: &mut H, byte: u8) {
+        assert!(
+            self.next_bit % 8 == 0,
+            "cannot add byte when not byte aligned"
+        );
+
+        let chunk_byte = self.next_bit / 8;
+        self.last_chunk_mut()[chunk_byte] = byte;
+        self.next_bit += 8;
+        if self.next_bit == Self::CHUNK_SIZE * 8 {
+            self.commit_last_chunk(hasher);
+        }
```

Comment:
I wonder if we should `panic!` if `next_bit` is ever larger than `CHUNK_SIZE * 8`.

Might not be worth checking but maybe it is

Reply:
I'm generally a fan of asserting on invariants and failing fast when violated, so added a couple assertions around this.

Reply:
My thinking was maybe leave it here for now and consider a broader refactor once the mutable store is functional.